plugins/batterystat.koplugin/main.lua

```diff
@@ -0,0 +1,273 @@
+    self.awake = Usage:new(self.settings:readSetting("awake"))
+    self.sleeping = Usage:new(self.settings:readSetting("sleeping"))
+
+    -- Note: these fields are not the "real" timestamp and battery usage, but
+    -- the unaccumulated values.
+    self.charging_state = State:new(self.settings:readSetting("charging_state"))
+    self.awake_state = State:new(self.settings:readSetting("awake_state"))
+    -- Whether the device was suspending before current timestamp.
+    self.was_suspending = false
+    -- Whether the device was charging before current timestamp.
+    self.was_charging = PowerD:isCharging()
+end
+
+function BatteryStat:onFlushSettings()
+    self.settings:replace({
+        charging = self.charging,
+        decharging = self.decharging,
+        awake = self.awake,
+        sleeping = self.sleeping,
+        charging_state = self.charging_state,
+        awake_state = self.awake_state,
+    })
+    self.settings:flush()
+end
+
+function BatteryStat:accumulate()
+    if self.was_suspending then
+        -- Suspending to awake.
+        self.sleeping:append(self.awake_state)
+    else
+        -- Awake to suspending, time between self.awake_state and now should belong to awake.
+        self.awake:append(self.awake_state)
+    end
+    if self.was_charging then
+        -- Decharging to charging.
+        self.charging:append(self.charging_state)
+    else
+        self.decharging:append(self.charging_state)
+    end
+    self.awake_state = State:new()
+    self.charging_state = State:new()
+end
+
+function BatteryStat:dumpOrLog(content)
+    local file = io.open(self.dump_file, "a")
+    if file then
+        file:write(content .. "\n")
+        file:close()
+    else
+        logger.warn("Failed to dump output ", content, " into ", self.dump_file )
+    end
+end
+
+function BatteryStat:debugOutput(event)
+    if self.debugging then
+        self:dumpOrLog(event .. " @ " .. State:new():toString() ..
+                       ", awake_state " .. self.awake_state:toString() ..
+                       ", charging_state " .. self.charging_state:toString())
+    end
+end
+
+function BatteryStat:onSuspend()
+    self:debugOutput("onSuspend")
+    self.was_suspending = false
+    self:accumulate()
+end
+
+function BatteryStat:onResume()
+    self:debugOutput("onResume")
+    self.was_suspending = true
+    self:accumulate()
+end
+
+function BatteryStat:onCharging()
+    self:debugOutput("onCharging")
+    self.was_charging = false
+    self:dumpToText()
+    self.charging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:onNotCharging()
+    self:debugOutput("onNotCharging")
+    self.was_charging = true
+    self:dumpToText()
+    self.decharging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:onCallback()
```

Comment:
I point it out because when I reach `onCallback` during code view, it's was clear for me what this function does and where it is called, I had to search function in the code. It would be better if we give this function a better name.

Yes, it's generally a good practice to avoid anonymous functions. In this case, `callback = function() end` already created the anonymous function, the size of a anonymous function doesn't really matter.

If you are concerned with indentation, you can rewrite it as the following form:

```lua
    local table_item = {
        text = _("Battery statistics"),
        callback = function()
	    BatteryStat:accumulate()
	    local kv_pairs = BatterStat:dump()
	    table.insert(kv_pairs, "----------------------------"})
	    table.insert(kv_pairs, {_("Historical records are dumped to"), ""})
	    table.insert(kv_pairs, {self.dump_file, ""})
	    UIManager:show(KeyValuePage:new{
		title = _("Battery statistics"),
		kv_pairs = kv_pairs,
	    })
        end,
    }
    table.insert(tab_item_table.plugins, table_item)
```